### ISSUE TITLE
Undefined behavior sanitizer throws an error in collection_notifier.cpp:51

### ISSUE THREAD

**Issue by kachalo** [realm/realm-swift#8245]
### How frequently does the bug occur?

Always

### Description

After modifying realm, Xcode always breaks at collection_notifer.cpp:51 due to Undefined Behavior Sanitizer breakpoint.

### Stacktrace & log output

```shell
Undefined Behavior
/Users/andriy/Library/Developer/Xcode/DerivedData/Subjects-ajxedkhgwredtogyrkhmfrjjfpea/SourcePackages/checkouts/realm-core/src/realm/object-store/impl/collection_notifier.cpp:51 Load of value 160, which is not a valid value for type 'bool'

Invalid bool load
#0	0x00000001100df8cf in realm::_impl::CollectionNotifier::get_modification_checker(realm::_impl::TransactionChangeInfo const&, realm::ConstTableRef) at /Users/andriy/Library/Developer/Xcode/DerivedData/Subjects-ajxedkhgwredtogyrkhmfrjjfpea/SourcePackages/checkouts/realm-core/src/realm/object-store/impl/collection_notifier.cpp:51
#1	0x0000000110167b77 in realm::_impl::ListNotifier::run() at /Users/andriy/Library/Developer/Xcode/DerivedData/Subjects-ajxedkhgwredtogyrkhmfrjjfpea/SourcePackages/checkouts/realm-core/src/realm/object-store/impl/list_notifier.cpp:99
#2	0x000000011018f427 in realm::_impl::RealmCoordinator::run_async_notifiers() at /Users/andriy/Library/Developer/Xcode/DerivedData/Subjects-ajxedkhgwredtogyrkhmfrjjfpea/SourcePackages/checkouts/realm-core/src/realm/object-store/impl/realm_coordinator.cpp:981
#3	0x000000011018cdc3 in realm::_impl::RealmCoordinator::on_change() at /Users/andriy/Library/Developer/Xcode/DerivedData/Subjects-ajxedkhgwredtogyrkhmfrjjfpea/SourcePackages/checkouts/realm-core/src/realm/object-store/impl/realm_coordinator.cpp:839
#4	0x000000011005686f in realm::_impl::ExternalCommitHelper::listen() at /Users/andriy/Library/Developer/Xcode/DerivedData/Subjects-ajxedkhgwredtogyrkhmfrjjfpea/SourcePackages/checkouts/realm-core/src/realm/object-store/impl/apple/external_commit_helper.cpp:234
#5	0x000000011005842b in realm::_impl::ExternalCommitHelper::ExternalCommitHelper(realm::_impl::RealmCoordinator&, realm::RealmConfig const&)::$_0::operator()() const at /Users/andriy/Library/Developer/Xcode/DerivedData/Subjects-ajxedkhgwredtogyrkhmfrjjfpea/SourcePackages/checkouts/realm-core/src/realm/object-store/impl/apple/external_commit_helper.cpp:175
#6	0x00000001100582ab in decltype(std::declval<realm::_impl::ExternalCommitHelper::ExternalCommitHelper(realm::_impl::RealmCoordinator&, realm::RealmConfig const&)::$_0>()()) std::__1::__invoke[abi:v15006]<realm::_impl::ExternalCommitHelper::ExternalCommitHelper(realm::_impl::RealmCoordinator&, realm::RealmConfig const&)::$_0>(realm::_impl::ExternalCommitHelper::ExternalCommitHelper(realm::_impl::RealmCoordinator&, realm::RealmConfig const&)::$_0&&) at /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.4.sdk/usr/include/c++/v1/__functional/invoke.h:394
#7	0x000000011005822f in void std::__1::__thread_execute[abi:v15006]<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, realm::_impl::ExternalCommitHelper::ExternalCommitHelper(realm::_impl::RealmCoordinator&, realm::RealmConfig const&)::$_0>(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, realm::_impl::ExternalCommitHelper::ExternalCommitHelper(realm::_impl::RealmCoordinator&, realm::RealmConfig const&)::$_0>&, std::__1::__tuple_indices<>) at /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.4.sdk/usr/include/c++/v1/thread:290
#8	0x000000011005755b in void* std::__1::__thread_proxy[abi:v15006]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, realm::_impl::ExternalCommitHelper::ExternalCommitHelper(realm::_impl::RealmCoordinator&, realm::RealmConfig const&)::$_0> >(void*) at /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.4.sdk/usr/include/c++/v1/thread:301
#9	0x00000001fea406b7 in _pthread_start ()
```

### Can you reproduce the bug?

Always

### Reproduction Steps

1. Enable Undefined Behavior Sanitizer in Scheme settings
2. Enable all runtime exceptions in Xcode
3. Run app that performs realm.write()

### Version

Realm 10.39.1, Realm Database 13.10.1

### What Atlas Services are you using?

Atlas Device Sync

### Are you using encryption?

No

### Platform OS and version(s)

iOS 16.4

### Build environment

Xcode version: Version 14.3 (14E222b)

**Comment by tgoyne** [comment]
It turns out our UBSan CI job isn't failing the build when UBSan reports errors, so a few have slipped in.

**Comment by tgoyne** [comment]
This specific one was already fixed by https://github.com/realm/realm-core/pull/6575 and just hasn't been pulled into a release yet.

**Comment by dianaafanador3** [comment]
@kachalo fix is already on our latest release [v10.40.0](https://github.com/realm/realm-swift/releases/tag/v10.40.0)
